### PR TITLE
Add sealunwrapper to ease OSS downgrades

### PR DESCRIPTION
**Pull request by jefferai**
Transparently downgrades shamir sealwrap protobufs, on active nodes only.

### CODE REVIEW

File: vault/sealunwrapper.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+		return nil, fmt.Errorf("cannot decode sealwrapped storage entry %s", entry.Key)
+	}
+	if atomic.LoadUint32(d.allowUnwraps) != 1 {
+		return &physical.Entry{
+			Key:   entry.Key,
+			Value: se.Ciphertext,
+		}, nil
+	}
+
+	entry = &physical.Entry{
```
Comment (chrishoffman) [inline vault/sealunwrapper.go]: You could create the entry above and just return it if not allowUnwraps.

File: vault/sealunwrapper.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
+package vault
```
[the remark this answers is not in the document]
Comment (jefferai) [reply vault/sealunwrapper.go]: I could do that, sure.

File: vault/sealunwrapper.go (hunk 1 above)
Comment (jefferai) [reply vault/sealunwrapper.go]: Done.

File: vault/sealunwrapper.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jefferai) [reply vault/sealunwrapper.go]: Done.

File: vault/sealunwrapper.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+func (d *transactionalSealUnwrapper) Transaction(ctx context.Context, txns []*physical.TxnEntry) error {
+	// Collect keys that need to be locked
+	var keys []string
+	for _, curr := range txns {
+		keys = append(keys, curr.Entry.Key)
+	}
+	// Lock the keys
+	for _, l := range locksutil.LocksForKeys(d.locks, keys) {
+		l.Lock()
```
Comment (vishalnayak) [inline vault/sealunwrapper.go]: Does this not cause a deadlock when 2 keys return the same lock?

Comment (chrishoffman) [reply vault/sealunwrapper.go]: LockForKeys is relatively new and will generate a list of unique locks.

**Pull request merged**